File: src/jsMain/kotlin/kingdom/lite/ui/KingdomApplicationV2.kt
package kingdom.lite.ui

import kotlinx.browser.window
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kingdom.lite.ui.components.*
import kotlin.js.Json
import kotlin.js.json

/**
 * External declaration for Foundry's ApplicationV2 class
 */
@JsModule("foundry.applications.api.ApplicationV2")
@JsNonModule
external abstract class ApplicationV2 {
    open val options: dynamic
    fun render(force: Boolean = definedExternally): dynamic
    fun close(options: dynamic = definedExternally): dynamic
    fun setPosition(position: dynamic = definedExternally): dynamic
}

/**
 * Creates a Foundry ApplicationV2-based Kingdom Sheet
 * Since Kotlin can't extend JS classes, we'll create and return a JS application instance
 */
fun createKingdomApplicationV2(): dynamic {
    // Simply use fallback to our custom implementation for now
    // The ApplicationV2 approach needs more complex JavaScript bridge
    console.log("Using custom Application implementation (ApplicationV2 bridge needs refinement)")
    return null
}

/**
 * Initialize the Kingdom Sheet HTML template
 */
fun initializeKingdomTemplate() {
    // Create the template HTML
    val template = buildString {
        append("""
            <div class="kingdom-container">
                <div class="kingdom-header">
                    ${ContentSelector { }.render()}
                </div>
                <div class="kingdom-body">
                    <div class="kingdom-sidebar">
                        ${KingdomStats.render()}
                    </div>
                    <div class="kingdom-main content">
                        ${ContentTurn().render()}
                    </div>
                </div>
            </div>
        """)
    }
    
    // Store it globally for the ApplicationV2 to use
    window.asDynamic().kingdomSheetV2Template = template
}

/**
 * Setup callbacks for the ApplicationV2
 */
fun setupKingdomCallbacks() {
    // Store components
    val contentSelector = ContentSelector { contentId ->
        switchContent(contentId)
    }
    
    val contentComponents = mapOf(
        "turn" to ContentTurn(),
        "settlements" to ContentSettlements,
        "factions" to ContentFactions,
        "modifiers" to ContentModifiers,
        "notes" to ContentNotes,
        "settings" to ContentSettings
    )
    
    var currentContent: ContentComponent? = contentComponents["turn"]
    
    // Setup render callback
    window.asDynamic().onKingdomSheetRendered = { element: dynamic ->
        console.log("Kingdom sheet rendered, element:", element)
        
        // Update the template with current content
        val mainArea = element.querySelector(".kingdom-main")
        if (mainArea != null) {
            mainArea.innerHTML = currentContent?.render() ?: ""
        }
    }
    
    // Setup listeners callback
    window.asDynamic().activateKingdomListeners = { html: dynamic ->
        val htmlElement = html as org.w3c.dom.HTMLElement
        
        console.log("Activating Kingdom listeners")
        
        // Attach component listeners
        contentSelector.attachListeners(htmlElement)
        KingdomStats.attachListeners(htmlElement)
        currentContent?.attachListeners(htmlElement)
        
        // Setup content switching
        val buttons = htmlElement.asDynamic().querySelectorAll(".content-selector button")
        val buttonsLength = buttons.length as? Int ?: 0
        for (i in 0 until buttonsLength) {
            val button = buttons.item(i)
            if (button != null) {
                button.onclick = { event: dynamic ->
                    val target = event.target
                    val contentId = target.getAttribute("data-content") ?: "turn"
                    
                    // Switch content
                    currentContent = contentComponents[contentId]
                    
                    // Update main area
                    val mainArea = htmlElement.querySelector(".kingdom-main") as? org.w3c.dom.HTMLElement
                    if (mainArea != null) {
                        mainArea.innerHTML = currentContent?.render() ?: ""
                        currentContent?.attachListeners(htmlElement)
                    }
                    
                    Unit
                }
            }
        }
    }
    
    window.asDynamic().switchContent = { contentId: String ->
        currentContent = contentComponents[contentId]
        
        val sheet = window.asDynamic().currentKingdomSheet
        if (sheet?.element != null) {
            val mainArea = sheet.element.querySelector(".kingdom-main")
            if (mainArea != null) {
                mainArea.innerHTML = currentContent?.render() ?: ""
                currentContent?.attachListeners(sheet.element as org.w3c.dom.HTMLElement)
            }
        }
    }
}

/**
 * Open the Kingdom Sheet using ApplicationV2
 */
fun openKingdomSheetV2() {
    console.log("Opening Kingdom Sheet V2 - falling back to custom implementation")
    
    // For now, fall back to our custom implementation
    // The ApplicationV2 approach would require a more complex JS bridge
    kotlinx.coroutines.GlobalScope.launch {
        val sheet = KingdomSheet()
        sheet.render(true)
    }
}

/**
 * Helper to switch content in the sheet
 */
private fun switchContent(contentId: String) {
    window.asDynamic().switchContent?.invoke(contentId)
}

File: src/ui/KingdomApplicationV2.ts
// Kingdom ApplicationV2 for PF2e Kingdom Lite
// Foundry VTT ApplicationV2 wrapper

import { ContentSelector } from './components/ContentSelector';
import { KingdomStats } from './components/KingdomStats';
import { ContentTurn } from './components/ContentTurn';
import { ContentSettlements } from './components/ContentSettlements';
import { ContentFactions } from './components/ContentFactions';
import { ContentModifiers } from './components/ContentModifiers';
import { ContentNotes } from './components/ContentNotes';
import { ContentSettings } from './components/ContentSettings';
import { ContentComponent } from './components/ContentComponent';

/**
 * External declaration for Foundry's ApplicationV2 class
 */
declare abstract class ApplicationV2 {
    open(options?: any): void;
    render(force?: boolean): any;
    close(options?: any): any;
    setPosition(position?: any): any;
}

/**
 * Creates a Foundry ApplicationV2-based Kingdom Sheet
 */
export function createKingdomApplicationV2(): any {
    // Simply use fallback to our custom implementation for now
    // The ApplicationV2 approach needs more complex JavaScript bridge
    console.log("Using custom Application implementation (ApplicationV2 bridge needs refinement)");
    return null;
}

/**
 * Initialize the Kingdom Sheet HTML template
 */
export function initializeKingdomTemplate(): string {
    // Create the template HTML
    const contentSelector = new ContentSelector((contentId) => {
        switchContent(contentId);
    });
    
    const template = `
        <div class="kingdom-container">
            <div class="kingdom-header">
                ${contentSelector.render()}
            </div>
            <div class="kingdom-body">
                <div class="kingdom-sidebar">
                    <!-- Kingdom stats will be rendered here -->
                </div>
                <div class="kingdom-main content">
                    ${new ContentTurn().render()}
                </div>
            </div>
        </div>
    `;
    
    // Store it globally for the ApplicationV2 to use
    (window as any).kingdomSheetV2Template = template;
    
    return template;
}

/**
 * Setup callbacks for the ApplicationV2
 */
export function setupKingdomCallbacks(): void {
    // Store components
    const contentSelector = new ContentSelector((contentId) => {
        switchContent(contentId);
    });
    
    const contentComponents = new Map<string, ContentComponent>([
        ["turn", new ContentTurn()],
        ["settlements", new ContentSettlements()],
        ["factions", new ContentFactions()],
        ["modifiers", new ContentModifiers()],
        ["notes", new ContentNotes()],
        ["settings", new ContentSettings()]
    ]);
    
    let currentContent: ContentComponent | null = contentComponents.get("turn") || null;
    
    // Setup render callback
    (window as any).onKingdomSheetRendered = (element: HTMLElement) => {
        console.log("Kingdom sheet rendered, element:", element);
        
        // Update the template with current content
        const mainArea = element.querySelector(".kingdom-main");
        if (mainArea) {
            mainArea.innerHTML = currentContent?.render() || "";
        }
    };
    
    // Setup listeners callback
    (window as any).activateKingdomListeners = (html: HTMLElement) => {
        const htmlElement = html as HTMLElement;
        
        console.log("Activating Kingdom listeners");
        
        // Attach component listeners
        contentSelector.attachListeners(htmlElement);
        // KingdomStats would be attached here if instantiated
        if (currentContent?.attachListeners) {
            currentContent.attachListeners(htmlElement);
        }
        
        // Setup content switching
        const buttons = htmlElement.querySelectorAll(".content-selector button");
        buttons.forEach((button) => {
            button.addEventListener('click', (event: Event) => {
                const target = event.target as HTMLElement;
                const contentId = target.getAttribute("data-content") || "turn";
                
                // Switch content
                currentContent = contentComponents.get(contentId) || null;
                
                // Update main area
                const mainArea = htmlElement.querySelector(".kingdom-main") as HTMLElement | null;
                if (mainArea) {
                    mainArea.innerHTML = currentContent?.render() || "";
                    if (currentContent?.attachListeners) {
                        currentContent.attachListeners(htmlElement);
                    }
                }
            });
        });
    };
    
    (window as any).switchContent = (contentId: string) => {
        currentContent = contentComponents.get(contentId) || null;
        
        const sheet = (window as any).currentKingdomSheet;
        if (sheet?.element) {
            const mainArea = sheet.element.querySelector(".kingdom-main");
            if (mainArea) {
                mainArea.innerHTML = currentContent?.render() || "";
                if (currentContent?.attachListeners) {
                    currentContent.attachListeners(sheet.element as HTMLElement);
                }
            }
        }
    };
}

/**
 * Open the Kingdom Sheet using ApplicationV2
 */
export function openKingdomSheetV2(): void {
    console.log("Opening Kingdom Sheet V2 - falling back to custom implementation");
    
    // For now, fall back to our custom implementation
    // The ApplicationV2 approach would require a more complex JS bridge
    import('./KingdomSheet').then(({ KingdomSheet }) => {
        const sheet = new KingdomSheet();
        sheet.render();
    });
}

/**
 * Helper to switch content in the sheet
 */
function switchContent(contentId: string): void {
    if ((window as any).switchContent) {
        (window as any).switchContent(contentId);
    }
}
